tests/Agency/StoreTestAPI.cpp

```diff
@@ -0,0 +1,977 @@
+  ASSERT_EQ(consensus::apply_ret_t::APPLIED, res.front());
+  assertEqual(read(R"([["/a"]])"), R"([{"a":2}])");
+  // in precondition & multiple
+  writeAndCheck(R"([[{"a":{"b":{"c":[1,2,3]},"e":[1,2]},"d":false}]])");
+  res = write(R"([[{"/b":2},{"/a/b/c":{"in":3}}]])");
+  ASSERT_EQ(consensus::apply_ret_t::APPLIED, res.front());
+  assertEqual(read(R"([["/b"]])"), R"([{"b":2}])");
+  res = write(R"([[{"/b":3},{"/a/e":{"in":3}}]])");
+  ASSERT_EQ(consensus::apply_ret_t::PRECONDITION_FAILED, res.front());
+  assertEqual(read(R"([["/b"]])"), R"([{"b":2}])");
+  res = write(R"([[{"/b":3},{"/a/e":{"in":3},"/a/b/c":{"in":3}}]])");
+  ASSERT_EQ(consensus::apply_ret_t::PRECONDITION_FAILED, res.front());
+  res = write(R"([[{"/b":3},{"/a/e":{"in":3},"/a/b/c":{"in":3}}]])");
+  ASSERT_EQ(consensus::apply_ret_t::PRECONDITION_FAILED, res.front());
+  res = write(R"([[{"/b":3},{"/a/b/c":{"in":3},"/a/e":{"in":3}}]])");
+  ASSERT_EQ(consensus::apply_ret_t::PRECONDITION_FAILED, res.front());
+  res = write(R"([[{"/b":3},{"/a/b/c":{"in":3},"/a/e":{"in":2}}]])");
+  ASSERT_EQ(consensus::apply_ret_t::APPLIED, res.front());
+  assertEqual(read(R"([["/b"]])"), R"([{"b":3}])");
+  
+  // Permute order of keys and objects within precondition
+  std::map<std::string,std::string> baz {
+    {"_id", "\"5a00203e4b660989b2ae5493\""},
+    {"index", "0"},
+    {"guid", "\"7a709cc2-1479-4079-a0a3-009cbe5674f4\""},
+    {"isActive", "true"},
+    {"balance", "\"$3,072.23\""},
+    {"picture", "\"http://placehold.it/32x32\""},
+    {"age", "21"},
+    {"eyeColor", "\"green\""},
+    {"name", R"({ "first": "Durham", "last": "Duke" })"},
+    {"tags", R"(["anim","et","id","do","est",1.0,-1024,1024])"}
+  };
+  std::string const baz_text = to_json_object(baz);
+  std::string const qux {R"(["3.14159265359",3.14159265359])"};
+  std::string const foo_value {"\"bar\""};
+  std::string const localObj = R"(
+      {"foo" : )"+ foo_value + R"(,
+        "baz" : )" + baz_text + R"(,
+        "qux" : )" + qux + R"(
+      })";
+
+  res = write(
+    std::string("[[") + localObj + R"(,
+      {
+        "baz":{"old": )" + baz_text + R"(},
+        "qux":)" + qux + R"(}]])");
+  ASSERT_EQ(consensus::apply_ret_t::PRECONDITION_FAILED, res.front());
+
+  writeAndCheck("[[" + localObj + "]]");
+  writeAndCheck(
+    "[[" + localObj + R"(, {"foo":)" + foo_value + R"(,"baz":{"old":)" + baz_text + R"(},"qux":)" + qux + "}]]");
+  writeAndCheck(
+    "[[" + localObj + R"(, {"baz":{"old":)" + baz_text + R"(},"foo":)" + foo_value + R"(,"qux":)" + qux + R"(}]])");
+  writeAndCheck(
+    "[[" + localObj + R"(, {"baz":{"old":)" + baz_text + R"(},"qux":)" + qux + R"(,"foo":)" + foo_value + R"(}]])");
+  writeAndCheck(
+    "[[" + localObj + R"(, {"qux":)" + qux + R"(,"baz":{"old":)" + baz_text + R"(},"foo":)" + foo_value + R"(}]])");
+
+  std::vector<std::string> localKeys;
+  std::transform(baz.begin(), baz.end(), std::back_inserter(localKeys), [](auto kv){ return kv.first; });
```

Comment:
Again, a matter of taste. I know this is considered "modern C++", but I prefer this style:

```
localKeys.reserve(baz.size));
for (auto const& p : baz ) {
  localKeys.push_back(p.first);
}
```

Yes, it is more lines, but it is easier to understand, IMHO.
No need to change, this is your style. Just wanted to mention it.